docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+# Lifted transformation
+
+⚠️ Advanced topic ⚠️
+
+This design note explains the underlying implementation of `flax.linen.transform` which enable JAX transformations inside `Module`.
+
+
+## Introduction
+
+JAX uses a functional API meaning that it only garantuees correct behavior when using functions without side effects.
+Typically these side effects are the result of mutating an object that lives outside of the function.
+
+The functional paradigm has some advantages like the ability to reason about state and stochasticity explicitly.
+A function output can only change when the input arguments change.
+Therefore, a function is garantueed to behave deterministicly.
+
+But pure functions offer another big advantage to JAX specifically: They enable functional transformations.
+For example `jax.vmap(f)` will vectorize a function `f`.
+Because `f` cannot have side-effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
+Would `f` draw the same or a different random number for each item in the vector?
+Would each item in the batch have its own counter or is the counter shared among the items?
+And in what order is the counter incremented if `f` is computed in parallel?
+The anser to all these questions is "it depends".
+The behavior is ambigious and the functional constraint elegantly avoids this problem.
+
+Flax introduces a safe way to have limited randomness and stateful variables in a JAX compatible form.
+The reason state in Flax is not problematic is because it is local. Inside a flax `Module` there are variables and PRNG sequences,
+but on the outside there are only JAX Arrays and PRNG keys.
+
+For most use cases Flax be used to define models in a stateful way.
+Becuase a `Module` behaves pure functional externally  we can utilize the full power of JAX with all of its transformations.
+There are however cases where we want to have the best of both worlds by using transformations and Module together.
+This design note explains how we extend JAX functional transformation to work on modules that have internal state and randomness.
+
+
+## Functionalization
+
+Before we jump into the details let's consider a simple example where we would like to use `vmap` inside a `Module`.
+
+First we define a simple MLP without any transformations:
+```
+import jax
+from jax import random, numpy as jnp
+from flax import linen as nn
+
+class MLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    h = nn.Dense(4, name='hidden')(xs)
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+```
+
+Now what if I want to have seperate MLP parameters for each item in `xs`?
+If this where "vanilla JAX" we could imagene writing something like `jax.vmap(apply_mlp)(mlp_params, xs)`.
```

Comment:
Suggestion:
- "imagine" (typo fix)
- proposal: "If we used "vanilla JAX" > "If this where "vanilla JAX" (or "this was" > "this where")

e.g.

```suggestion
If this was "vanilla JAX", we could imagine writing something like `jax.vmap(apply_mlp)(mlp_params, xs)`.
```